core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -858,10 +925,10 @@ class Partition(val topicPartition: TopicPartition,
         case Some(leaderLog) =>
           val outOfSyncReplicaIds = getOutOfSyncReplicas(replicaLagTimeMaxMs)
           if (outOfSyncReplicaIds.nonEmpty) {
-            val newInSyncReplicaIds = inSyncReplicaIds -- outOfSyncReplicaIds
+            val newInSyncReplicaIds = isrState.isr -- outOfSyncReplicaIds
```

Comment:
Seems like we do not have a check for inflight AlterIsr after the write lock has been acquired.